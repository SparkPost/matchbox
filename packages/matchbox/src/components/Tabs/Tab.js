import React, { Component } from 'react';
import classnames from 'classnames';

import { UnstyledLink } from '../UnstyledLink';

import styles from './Tabs.module.scss';

class Tab extends Component {
  static displayName = 'Tab';

  handleClick = (event) => {
    const { index, onClick } = this.props;

    if (onClick) {
      return onClick(event, index);
    }
  }

  render() {
    const { index, content, selected, fittedTab, ...rest } = this.props;
    const classes = classnames(
      styles.Tab,
      selected === index && styles.selected,
      fittedTab && styles.fittedTab
    );

    return (
      <UnstyledLink className={classes} {...rest} onClick={this.handleClick}>
        {content}
      </UnstyledLink>
    );
  }
}

export default Tab;
