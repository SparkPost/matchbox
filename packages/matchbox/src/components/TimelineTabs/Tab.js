import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { buttonReset, focusOutline } from '../../styles/helpers';
import { Box } from '../Box';

const Container = styled.button`
  ${buttonReset}
  position: relative;
  cursor: pointer;
  width: 100%;
  outline: none;

  svg circle {
    transition: ${({ theme }) => theme.motion.duration.fast};
  }

  svg [data-id='mb-tab-line'] {
    stroke: ${({ theme }) => theme.colors.gray[700]};
  }

  svg [data-id='mb-tab-outer'] {
    fill: ${({ theme }) => theme.colors.white};
    stroke: ${props =>
      props.selected ? props.theme.colors.blue[700] : props.theme.colors.gray[700]};
  }

  svg [data-id='mb-tab-inner'] {
    fill: ${props => (props.selected ? props.theme.colors.blue[700] : 'transparent')};
  }

  &:hover {
    svg [data-id='mb-tab-outer'] {
      stroke: ${({ theme, selected }) =>
        !selected ? theme.colors.gray[800] : theme.colors.blue[700]};
    }
    svg [data-id='mb-tab-inner'] {
      fill: ${({ theme, selected }) =>
        !selected ? theme.colors.gray[400] : theme.colors.blue[700]};
    }
  }

  ${'' /* &:focus svg [data-id='mb-tab-outer'] {
    stroke-width: 2;
  } */}

  &:last-child {
    svg line {
      display: none;
    }
  }
`;

const FocusSvg = styled.div`
  width: 24px;
  height: 24px;

  ${Container}:focus-visible & {
    ${focusOutline({ modifier: '&', offset: '1px', radius: '50%' })}
  }
`;

const Tab = React.forwardRef(function Tab(props, userRef) {
  const {
    as = 'button',
    children,
    onClick,
    handleParentSelect,
    selected,
    to,
    href,
    ...rest
  } = props;

  function handleClick(e) {
    if (onClick) {
      onClick(e);
    }
    handleParentSelect();
  }

  return (
    <Container
      aria-controls={rest['aria-controls']}
      aria-selected={selected}
      as={as}
      href={href}
      onClick={handleClick}
      ref={userRef}
      role="tab"
      selected={selected}
      tabIndex={selected ? '0' : '-1'}
      to={to}
      type="button"
    >
      <Box display="grid" gridTemplateColumns="30px 1fr" gridGap="200">
        <Box position="relative" height="100%" gridColumn="0/1">
          <Box
            position="absolute"
            width="2px"
            height="100%"
            top="2px"
            left="35%"
            bg="red"
            zIndex="below"
            overflow="hidden"
          >
            <svg width="5px" height="200px">
              <line
                data-id="mb-tab-line"
                x1="1"
                y1="0"
                x2="1"
                y2="100"
                stroke="black"
                strokeWidth="1"
                strokeDasharray="3 2"
              />
            </svg>
          </Box>
          <FocusSvg>
            <svg width="25px" height="25px">
              <circle
                data-id="mb-tab-outer"
                cx="12"
                cy="12"
                r="11"
                fill="transparent"
                strokeWidth="1"
              />
              <circle data-id="mb-tab-inner" cx="12" cy="12" r="5" />
            </svg>
          </FocusSvg>
        </Box>
        <Box textAlign="left" minHeight="800">
          {children}
        </Box>
      </Box>
    </Container>
  );
});

Tab.propTypes = {
  'aria-controls': PropTypes.string,
  as: PropTypes.elementType,
  children: PropTypes.node,
  handleParentSelect: PropTypes.func,
  href: PropTypes.string,
  onClick: PropTypes.func,
  selected: PropTypes.bool,
  to: PropTypes.string,
};

export default Tab;
