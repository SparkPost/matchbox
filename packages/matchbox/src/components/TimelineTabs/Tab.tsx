import React from 'react';
import styled from 'styled-components';
import { buttonReset, focusOutline } from '../../styles/helpers';
import { Box } from '../Box';
import type * as Polymorphic from '../../helpers/types';

const Container = styled.button<{ $selected?: boolean }>`
  ${buttonReset}
  position: relative;
  cursor: pointer;
  width: 100%;
  outline: none;

  svg circle {
    transition: ${({ theme }) => theme.motion.duration.fast};
  }

  svg [data-id='mb-tab-line'] {
    stroke: ${({ theme }) => theme.colors.gray[700]};
  }

  svg [data-id='mb-tab-outer'] {
    fill: ${({ theme }) => theme.colors.white};
    stroke: ${(props) =>
      props.$selected ? props.theme.colors.blue[700] : props.theme.colors.gray[700]};
  }

  svg [data-id='mb-tab-inner'] {
    fill: ${(props) => (props.$selected ? props.theme.colors.blue[700] : 'transparent')};
  }

  &:hover {
    svg [data-id='mb-tab-outer'] {
      stroke: ${({ theme, $selected }) =>
        !$selected ? theme.colors.gray[800] : theme.colors.blue[700]};
    }
    svg [data-id='mb-tab-inner'] {
      fill: ${({ theme, $selected }) =>
        !$selected ? theme.colors.gray[400] : theme.colors.blue[700]};
    }
  }

  &:last-child {
    svg line {
      display: none;
    }
  }
`;

const FocusSvg = styled.div`
  width: 24px;
  height: 24px;

  ${Container}:focus & {
    ${focusOutline({ modifier: '&', offset: '1px', radius: '50%' })}
  }

  ${Container}:focus:not(:focus-visible) &:after {
    box-shadow: none;
  }

  ${Container}:focus-visible & {
    ${focusOutline({ modifier: '&', offset: '1px', radius: '50%' })}
  }
`;

export type TimelineTabsTabProps = {
  'data-id'?: string;
  handleParentSelect?: () => void;
  selected?: boolean;
};

type PolymorphicTab = Polymorphic.ForwardRefComponent<'button', TimelineTabsTabProps>;

const Tab = React.forwardRef(function Tab(props, userRef) {
  const {
    as = 'button',
    'aria-controls': ariaControls,
    children,
    'data-id': dataId,
    id,
    handleParentSelect,
    onClick,
    selected,
    ...rest
  } = props;

  function handleClick(e) {
    if (onClick) {
      onClick(e);
    }
    handleParentSelect();
  }

  return (
    <Container
      aria-controls={ariaControls}
      aria-selected={selected}
      as={as}
      data-id={dataId}
      id={id}
      onClick={handleClick}
      ref={userRef}
      role="tab"
      $selected={selected}
      tabIndex={selected ? 0 : -1}
      type="button"
      {...rest}
    >
      <Box display="grid" gridTemplateColumns="30px 1fr" gridGap="200">
        <Box position="relative" height="100%" gridColumn="0/1">
          <Box
            position="absolute"
            width="2px"
            height="100%"
            top="2px"
            left="35%"
            bg="red"
            zIndex="below"
            overflow="hidden"
          >
            <svg width="5px" height="200px">
              <line
                data-id="mb-tab-line"
                x1="1"
                y1="0"
                x2="1"
                y2="100"
                stroke="black"
                strokeWidth="1"
                strokeDasharray="3 2"
              />
            </svg>
          </Box>
          <FocusSvg>
            <svg width="25px" height="25px">
              <circle
                data-id="mb-tab-outer"
                cx="12"
                cy="12"
                r="11"
                fill="transparent"
                strokeWidth="1"
              />
              <circle data-id="mb-tab-inner" cx="12" cy="12" r="5" />
            </svg>
          </FocusSvg>
        </Box>
        <Box textAlign="left" minHeight="800" pb="300">
          {children}
        </Box>
      </Box>
    </Container>
  );
}) as PolymorphicTab;

export default Tab;
