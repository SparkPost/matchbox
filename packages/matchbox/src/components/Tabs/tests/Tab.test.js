import React from 'react';
import { shallow } from 'enzyme';

import Tab from '../Tab';

describe('Tab', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      content: 'Example',
      index: 99
    };
    wrapper = shallow(<Tab {...props} />);
  });

  it('should render a tab', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should return index to click handler', () => {
    const fakeEvent = Symbol;
    const onClick = jest.fn();

    wrapper.setProps({ onClick });
    wrapper.simulate('click', fakeEvent);

    expect(onClick).toHaveBeenCalledWith(fakeEvent, 99);
  });
});
