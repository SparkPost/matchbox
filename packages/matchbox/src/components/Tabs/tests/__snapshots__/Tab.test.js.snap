// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab should render a tab 1`] = `
<UnstyledLink
  className="Tab"
  onClick={[Function]}
>
  Example
</UnstyledLink>
`;
